tscv/_split.py

```diff
@@ -371,6 +372,128 @@ def get_n_splits(self, X=None, y=None, groups=None):
+                             % (N, type(N)))
+        N = int(N)
+
+        if not isinstance(k, numbers.Integral):
+            raise ValueError('The number of test splits must be of Integral '
+                             'type. %s of type %s was passed.'
+                             % (k, type(k)))
+        k = int(k)
+
+        if N <= 1:
+            raise ValueError(
+                "Combinatorial k-fold cross-validation requires at least two"
+                " groups by setting N=2 or more,"
+                " got N={0}.".format(N))
+
+        if k < 1:
+            raise ValueError(
+                "Combinatorial k-fold cross-validation requires at least one"
+                " test split by setting k=1 or more,"
+                " got k={0}.".format(k))
+
+        super().__init__(gap_before, gap_after)
+        self.n_groups = N
+        self.test_splits = k
+
+    def split(self, X, y=None, groups=None):
+        n_samples = _num_samples(X)
+        n_splits = self.n_groups
+        gap_before, gap_after = self.gap_before, self.gap_after
+        if n_splits > n_samples:
+            raise ValueError(
+                ("Cannot have number of splits n_splits={0} greater"
+                 " than the number of samples: n_samples={1}.")
+                .format(self.n_groups, n_samples))
+        self.indexes = np.arange(n_samples)
+        splits = [(split[0], split[-1] + 1)
+                  for split
+                  in np.array_split(self.indexes, self.n_groups)]
+        splits_combinations = list(combinations(splits, self.test_splits))
+        for splits_combination in splits_combinations:
+            test_indexes = np.empty(0)
+            train_indexes = self.indexes
+            for start, stop in splits_combination:
+                test_indexes = np.union1d(
+                    test_indexes, self.indexes[start:stop]).astype(int)
+                begin = max(0, start - gap_before)
+                end = min(n_samples, stop + gap_after)
+                train_indexes = np.intersect1d(train_indexes,
+                    np.setdiff1d(self.indexes, self.indexes[begin:end]))
+                if len(train_indexes) <= 0:
+                    raise ValueError("Not enough training samples available")
+            yield train_indexes, test_indexes
+
+    def get_n_splits(self, X=None, y=None, groups=None):
+        """Returns the number of splitting iterations in the cross-validator
+
+        Parameters
+        ----------
+        X : object
+            Always ignored, exists for compatibility.
+
+        y : object
+            Always ignored, exists for compatibility.
+
+        groups : object
+            Always ignored, exists for compatibility.
+
+        Returns
+        -------
+        n_splits : int
+            Returns the number of splitting iterations in the cross-validator.
+        """
+        return len(list(combinations(range(self.n_groups), self.test_splits)))
```

Comment:
Use the combination number to generate the result directly rather than instantiating all combinations.